### PR TITLE
Add Support for DingoDB as vectorstore

### PR DESCRIPTION
**Pull request by HeChangHaoGary**
Description:
[DingoDB](https://dingodb.readthedocs.io/en/latest/) is a distributed multi-mode vector database, which combines the characteristics of data lakes and vector databases. It has real-time low-latency processing capabilities to achieve rapid insight and response, and can efficiently conduct instant analysis and process multi-modal data.

### CODE REVIEW

File: libs/langchain/langchain/retrievers/dingo.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+from langchain.schema import BaseRetriever, Document
+from langchain.vectorstores.dingo import Dingo
+
+
+class DingoRetriever(BaseRetriever):
+    """Retriever that uses the Dingo API"""
+
+    embeddings: Embeddings
+    """Embeddings model to use."""
+    """description"""
```
Comment (hwchase17) [inline libs/langchain/langchain/retrievers/dingo.py:17]: stray comment

File: libs/langchain/langchain/retrievers/dingo.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+"""Dingo Retriever"""
```
Comment (hwchase17) [inline libs/langchain/langchain/retrievers/dingo.py:1]: what does this add over the vectorstore? if it doesnt add anything, may be best to delete

**Review by hwchase17 — COMMENTED** [review]
@baskaryan are these changes to the docs_skeleton and api_reference needed?

File: libs/langchain/langchain/retrievers/dingo.py (hunk 2 above)
Comment (HeChangHaoGary) [reply libs/langchain/langchain/retrievers/dingo.py]: > what does this add over the vectorstore? if it doesnt add anything, may be best to delete

I hava Fixed， removed this description

File: libs/langchain/langchain/retrievers/dingo.py (hunk 1 above)
Comment (HeChangHaoGary) [reply libs/langchain/langchain/retrievers/dingo.py]: > stray comment
I hava Fixed， removed this """description"""

File: libs/langchain/langchain/retrievers/dingo.py (hunk 2 above)
Comment (baskaryan) [reply libs/langchain/langchain/retrievers/dingo.py]: i think what harrison means is that you can already take any vector store and turn it into a retriever pretty easily 
```python
vectorstore = Dingo(...)
retriever = vectorstore.as_retriever()
```

and that retriever will under the hood use the vectorstore to perform similarity search. so probably no need to manually implement a DingoRetriever class as well, can delete

**Comment by baskaryan** [comment]
> @baskaryan are these changes to the docs_skeleton and api_reference needed?

nope, no need for redirects since this is a new page and api reference is auto generated

File: libs/langchain/langchain/retrievers/dingo.py (hunk 2 above)
Comment (HeChangHaoGary) [reply libs/langchain/langchain/retrievers/dingo.py]: > and that retriever will under the hood use the vectorstore to perform similarity search. so probably no need to manually implement a DingoRetriever class as well, can delete

Thanks for reminding me, I have deleted

File: libs/langchain/langchain/vectorstores/dingo.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+            metadatas_list.append(metadata)
+            embeds.append(emb)
+            
+
+        # upsert to Dingo
+        for i in range(0, len(texts), batch_size):
+            j = i + batch_size
+            self._client.vector_add(self._index_name, metadatas_list[i:j], embeds[i:j], ids[i:j])
+
+        return ids
```
Comment (baskaryan) [inline libs/langchain/langchain/vectorstores/dingo.py:114]: this is expected to be a list of strings

Comment (HeChangHaoGary) [reply libs/langchain/langchain/vectorstores/dingo.py:111]: > 

Can I change the return type to List[int]?  Because our incoming id is int type

Comment (baskaryan) [reply libs/langchain/langchain/vectorstores/dingo.py:111]: we could use ints for actually uploading but return
```python
return [str(id) for id in ids]
```

Comment (HeChangHaoGary) [reply libs/langchain/langchain/vectorstores/dingo.py:111]: > we could use ints for actually uploading but return
> 
> ```python
> return [str(id) for id in ids]
> ```

Ok, I try to modify it back to list[str], the python SDK will automatically convert it with pydantic

**Comment by baskaryan** [comment]
if you want you can merge #9079 into this branch and that should fix

**Comment by baskaryan** [comment]
landed in #9079, thanks @HeChangHaoGary!

**Pull request closed without merging**